### PR TITLE
Add backward check for test_memory_format

### PR DESCRIPTION
**Pull request by CaoE**
Add backward check for test_memory_format.

### CODE REVIEW

**Review by jgong5 — COMMENTED** [review]
For the cases where the backward memory format checks are disabled, are they issues that need to be fixed?

**Comment by CaoE** [comment]
> For the cases where the backward memory format checks are disabled, are they issues that need to be fixed?

Will collect such issue. If there are no related issues, I will create corresponding github issues.

**Comment by CaoE** [comment]
@mikaylagawarecki  For test breakings on MPS and  CUDA, I created corresponding issues https://github.com/pytorch/pytorch/issues/107214, https://github.com/pytorch/pytorch/issues/107199, and https://github.com/pytorch/pytorch/issues/107201.

I have no mac machine to reproduce test breaking on MPS, so I'm waiting replies if it is real issue https://github.com/pytorch/pytorch/issues/107214. Then this PR should be ready.
From my perspective, the code is ready to be reviewed. Do you mind review this draft PR first ?

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -2524,6 +2526,8 @@ def module_error_inputs_torch_nn_Pad3d(module_info, device, dtype, requires_grad
                ),
     ModuleInfo(torch.nn.AdaptiveAvgPool2d,
                gradcheck_nondet_tol=GRADCHECK_NONDET_TOL,
+               # Fails on backward check if output size is 1x1
+               gradcheck_memformat=False,
```
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_modules.py:2530]: Appreciate the detailed comments for the reasons why each of these are marked False! Instead of `gradcheck_memformat` could we xfail these? That way if the failing tests are fixed we will have signal in CI

File: test/test_modules.py [hunk 2]
```diff
@@ -691,14 +722,51 @@ def inner_check_out_mem_format(output):
...
+                        diff_outputs = tuple(t for t in outputs if _req_grad(t))
+                        if gradcheck_memformat and training and len(diff_outputs) > 0:
+                            params = tuple(p for p in m.parameters())
+                            diff_inputs = tuple(
+                                t
+                                for t in pytree.tree_flatten((args, kwargs, params))[0]
+                                if _req_grad(t)
+                            )
+                            grad_outputs = tuple(
+                                torch.rand_like(t)
```
Comment (mikaylagawarecki) [inline test/test_modules.py:744]: nit: 
```suggestion
                                torch.empty_like(t)
```

File: torch/testing/_internal/common_modules.py (hunk 1 above)
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_modules.py:2530]: Really appreciate the detailed comments here regarding why each of these is set to False! 
 
nit: Separately, is there any way we could make this an xfail? I'm hoping that if these are fixed, whoever sends the PR fixing will get signal to un-xfail these tests. I worry that if we make this a flag, re-enabling the tests might fall through the cracks.

I'm thinking maybe something to the effect of 
```
DecorateInfo(
     unittest.expectedFailure, 'TestModule',
     'test_memory_format',
     active_if=lambda p: p['training']
 )
```
This seems reasonable to me because the non-backward version will still be tested if training=False and we can change the check [here](https://github.com/pytorch/pytorch/blob/499aaf6cd790ba1008fbacabf8b0fd26d93ca289/test/test_modules.py#L693) to `if training and len(ref_diff_outputs) > 0`

Let me know your thoughts on this!

**Review by mikaylagawarecki — COMMENTED** [review]
@CaoE Thank you very much for your hard work with this PR as well as filing issues for the failing tests!!

Regarding the failing mac test, I think the issue is sufficient and it is ok to skip this test when landing this PR.

Separately, what do you think of removing the corresponding tests in test_nn.py?

**Comment by mikaylagawarecki** [comment]
One more thing -- do you think it makes sense to extend this to test that the gradient for the channels_last params/buffers of the model have gradients of the correct memory format as well. Could be a followup but I was looking into https://github.com/pytorch/pytorch/issues/107199 and it made me wonder whether we might have silent correctness issues for memory format of params/buffers as well?

File: test/test_modules.py [hunk 3]
```diff
@@ -676,6 +686,27 @@ def inner_check_out_mem_format(output):
                 # === Get output in (contiguous, contiguous) configuration. ===
                 args, kwargs = module_input.forward_input.args, module_input.forward_input.kwargs
                 desired_outputs = m(*args, **kwargs)
+                if isinstance(desired_outputs, torch.Tensor):
+                    desired_outputs = (desired_outputs,)
+                # === Do backward pass. ===
+                ref_diff_outputs = tuple(t for t in desired_outputs if _req_grad(t))
```
[the remark this answers is not in the document]
Comment (CaoE) [reply test/test_modules.py]: Used pytree.tree_flatten instead.

File: test/test_modules.py [hunk 4]
```diff
@@ -676,6 +686,27 @@ def inner_check_out_mem_format(output):
...
+                    )
+                    ref_grad_outputs = tuple(
+                        torch.rand_like(t)
+                        for t in ref_diff_outputs
+                    )
+                    ref_grad_inputs = torch.autograd.grad(
+                        ref_diff_outputs,
+                        ref_diff_inputs,
+                        grad_outputs=ref_grad_outputs,
+                        allow_unused=True,
```
[the remark this answers is not in the document]
Comment (CaoE) [reply test/test_modules.py:704]: Thanks for your comments. I removed `allow_unused=True`.

File: test/test_modules.py [hunk 5]
```diff
@@ -691,14 +722,51 @@ def inner_check_out_mem_format(output):
...
+                            grad_inputs = torch.autograd.grad(
+                                diff_outputs,
+                                diff_inputs,
+                                grad_outputs=grad_outputs,
+                                allow_unused=True,
+                            )
+
+                            if (
+                                input_mem_format != torch.contiguous_format
+                                or module_mem_formats != torch.contiguous_format
```
[the remark this answers is not in the document]
Comment (CaoE) [reply test/test_modules.py]: Fixed as suggested. I also found https://github.com/pytorch/pytorch/pull/106104/files#diff-e449b2c5758c5d63d7ea8ff8f7d918dd4071ab45bd82d843fc34b522b4d143aeL696 has similar code.

File: test/test_modules.py [hunk 6]
```diff
@@ -691,14 +722,51 @@ def inner_check_out_mem_format(output):
...
+                                if _req_grad(t)
+                            )
+                            grad_outputs = tuple(
+                                torch.rand_like(t)
+                                for t in diff_outputs
+                            )
+                            grad_outputs = tuple(
+                                t1.copy_(t2)
+                                for (t1, t2) in zip(grad_outputs, ref_grad_outputs)
+                            )
```
[the remark this answers is not in the document]
Comment (CaoE) [reply test/test_modules.py]: It is cleaner. Fixed as suggested.

File: torch/testing/_internal/common_modules.py (hunk 1 above)
Comment (CaoE) [reply torch/testing/_internal/common_modules.py]: Thanks for your suggestion! It‘s a good idea. I just found there is `active_if` by which `DecorateInfo` can be activated according to input parameters of the test.
Modified as suggested.

**Comment by CaoE** [comment]
> Separately, what do you think of removing the corresponding tests in test_nn.py?

I think most memory_format test in test_nn.py can be removed except for some edge cases (specific shapes, etc.)

**Comment by CaoE** [comment]
> I was looking into https://github.com/pytorch/pytorch/issues/107199 and the complexity of the code paths/amount of branching made me wonder whether we might have silent correctness issues for memory format of gradients of params/buffers as well?

From my perspective,  I think some kernels (for forward and backward on different device) do not implement channels last support and they may return non-channels last results for channels last inputs. This break channels last's propagation but should not cause any computing error (if there is any, it's a bug). Developers should add CL support as much as possible to avoid breaking the propagation of CL and achieve better performance.

File: test/test_modules.py [hunk 7]
```diff
@@ -693,12 +720,43 @@ def inner_check_out_mem_format(output):
                         outputs = m(*args, **kwargs)
 
                         # === Compare outputs to (contiguous, contiguous) output. ===
-                        if input_mem_format != torch.contiguous_format or module_mem_formats != torch.contiguous_format:
+                        if input_mem_format != torch.contiguous_format or module_mem_format != torch.contiguous_format:
```
Comment (mikaylagawarecki) [inline test/test_modules.py:723]: thanks for catching this!

File: torch/testing/_internal/common_modules.py [hunk 8]
```diff
@@ -2564,7 +2571,16 @@ def module_error_inputs_torch_nn_Pad3d(module_info, device, dtype, requires_grad
...
+                   DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),
+                   # The difference between channels last backward and
+                   # channels first backward of AvgPool2d on CUDA is too large
+                   # See https://github.com/pytorch/pytorch/issues/107201
+                   DecorateInfo(
+                       unittest.expectedFailure,
+                       'TestModule',
+                       'test_memory_format',
+                       active_if=lambda p: p['training'] and p['device'] == 'cuda',
+                   ),),
```
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_modules.py:2583]: I think `p['device'] == 'cuda'` here might not work and is causing the CI failure because the device will actually be something like `'cuda:0'` (verified this locally). This should work instead

```suggestion
                   DecorateInfo(
                       unittest.expectedFailure,
                       'TestModule',
                       'test_memory_format',
                       active_if=lambda p: p['training'],
                       device_type='cuda',
                   ),)
```

File: torch/testing/_internal/common_modules.py [hunk 9]
```diff
@@ -2635,7 +2659,15 @@ def module_error_inputs_torch_nn_Pad3d(module_info, device, dtype, requires_grad
                        unittest.expectedFailure, 'TestEagerFusionModuleInfo',
                        'test_aot_autograd_module_exhaustive',
                        active_if=lambda p: p['training']
-                   ))
+                   ),
+                   # native_batch_norm_backward on cuda does not support channels last.
+                   # see https://github.com/pytorch/pytorch/issues/107199
+                   DecorateInfo(
```
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_modules.py:2665]: I have a fix for this https://github.com/pytorch/pytorch/pull/107270

**Review by mikaylagawarecki — APPROVED** [review]
LGTM! Please fix the xfails with  `active_if=lambda p: p['training'] and p['device'] == 'cuda'` per the comment and additionally xfail the MPS HardSwish test

**Comment by mikaylagawarecki** [comment]
@CaoE thanks again for adding this test. Following up,

1)  I will do an in depth review of https://github.com/pytorch/pytorch/pull/97775 (Could you update that to make sure this test is running on that PR)
2) Would you be interested in removing the related `test_nn.py` tests?

File: torch/testing/_internal/common_modules.py [hunk 10]
```diff
@@ -2545,7 +2545,14 @@ def module_error_inputs_torch_nn_Pad3d(module_info, device, dtype, requires_grad
                module_inputs_func=module_inputs_torch_nn_AdaptiveAvgPool2d,
                skips=(
                    # Fails on MPS backend if input/output sizes are not divisible
-                   DecorateInfo(skipMPS),)
+                   DecorateInfo(skipMPS),
+                   # Fails on backward check if output size is 1x1
+                   DecorateInfo(
```
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_modules.py:2550]: Seems like the issue is that this gives an unexpected success when run using inductor. Do you know what the issue is here for 1x1 outputs for eager, would it be possible to fix it? 

Otherwise I am okay with skipping the test and filing an issue

Comment (CaoE) [reply torch/testing/_internal/common_modules.py:2593]: For 1x1 outputs, mean is applied on CPU instead of AdaptiveAvgPool2d so the grad will be channels first . Firstly, I added expectedFailure for cpu but it is still failed on CUDA. I will check this further.

Comment (CaoE) [reply torch/testing/_internal/common_modules.py:2593]: For unexpected success when run using inductor,  please see https://github.com/pytorch/pytorch/issues/107861

**Comment by CaoE** [comment]
> @CaoE thanks again for adding this test. Following up,
> 
> 1. I will do an in depth review of [add channel last 3d support for maxpool3d on CPU #97775](https://github.com/pytorch/pytorch/pull/97775) (Could you update that to make sure this test is running on that PR)
> 2. Would you be interested in removing the related `test_nn.py` tests?

Sorry for slow reply. I'm occupied by some urgent tasks recently. 
I may not have much time to do this in short time, but I can gradually remove these tests in later PRs.

**Comment by CaoE** [comment]
@mikaylagawarecki Seems https://github.com/pytorch/pytorch/actions/runs/5972147266/job/16212838253 and https://github.com/pytorch/pytorch/actions/runs/5972147266/job/16212838606 are not related to this PR. Can we skip them and merge this PR ?

File: torch/testing/_internal/common_modules.py (hunk 10 above)
Comment (mikaylagawarecki) [reply torch/testing/_internal/common_modules.py:2593]: Hm how come this doesn't give unexpected success in CI on this PR anymore?

Comment (mikaylagawarecki) [reply torch/testing/_internal/common_modules.py:2593]: Oh thank you! I couldn't see the diff because it was force pushed, merging now
